### PR TITLE
No rendering after drop

### LINKED ISSUE #6505: Using wasm `drop` method results in warnings flood
This is an issue to fix leftovers after implementing `wasm.drop` method. Original problem was rised during QAing https://github.com/enso-org/enso/pull/6365#issuecomment-1521870772

Steps to reproduce:
1. Open Enso with `-debug.dev-tools` flag
2. Navigate to console pane
3. Type `ensoglApp.wasm.drop()`
4. Warnings will flood both console and terminal

https://drive.google.com/file/d/1Oy3-D5JGYy4PbJhYAXYXB3Ouuks_zL45/view?usp=share_link

### LINKED ISSUE #6976: Once you get an error message you can't get rid of it
An error caused by a visualization timeout was shown (great to see) but it is impossible to get rid of this message so remains there until you reload Enso.

Need a dismiss button (or to dismiss after a period of time).

### PR DESCRIPTION
**Pull request by Procrat**
### Description

There were various memory leaks which caused the scene to stay around which tries to render after the GUI has been dropped.

There were 1500+ leaked objects. The first few commits by @Procrat tackle the actual leaks. The last commit of him tries to reduce the cloning of `Application` across the codebase, so debugging leaking references will be easier. @farmaazon then added fixes for last leaks and cleaned up the code.

Additionally, the message about visualization failing to attach is displayed in pop-up instead of status bar, so it will disappear after some time. Fixes #6976 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- ~~[ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui/view/graph-editor/src/component/node/input/widget/single_choice.rs [hunk 1]
```diff
@@ -400,6 +400,7 @@ fn entry_for_current_value(
 /// is minimal, as the actual dropdown view is not created.
 #[derive(Debug)]
 struct LazyDropdown {
+    // Required for lazy initialization
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/widget/single_choice.rs:403]: dot missing

File: app/gui/view/graph-editor/src/lib.rs [hunk 2]
```diff
@@ -1802,6 +1805,7 @@ impl GraphEditorModelWithNetwork {
 #[allow(missing_docs)] // FIXME[everyone] Public-facing API should be documented.
 pub struct GraphEditorModel {
     pub display_object: display::object::Instance,
+    // Required for dynamically creating nodes and edges
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/lib.rs]: dot missing

File: lib/rust/ensogl/component/list-view/src/entry/list.rs [hunk 3]
```diff
@@ -66,6 +66,7 @@ pub type List<E> = ListData<E, <E as Entry>::Params>;
 #[derivative(Clone(bound = ""))]
 #[clone_ref(bound = "E:CloneRef")]
 pub struct ListData<E, P> {
+    // Required for dynamically creating new entries
```
Comment (wdanilo) [inline lib/rust/ensogl/component/list-view/src/entry/list.rs]: dot missing

File: lib/rust/ensogl/component/sequence-diagram/src/lib.rs [hunk 4]
```diff
@@ -89,6 +89,7 @@ impl component::Frp<Model> for Frp {
 /// Internal model of the SequenceDiagram.
 #[derive(Clone, CloneRef, Debug)]
 pub struct Model {
+    // Required for dynamically creating new lines
```
Comment (wdanilo) [inline lib/rust/ensogl/component/sequence-diagram/src/lib.rs]: missing dot

**Review by wdanilo — APPROVED** [review]
These changes look OK, but I'm missing info how this PR really works. Even if we are cloning world and scene in different places, all of these objects should get dropped if we drop the main app instance. Even if FRP closures keep reference to them, these networks should be dropped, so how this PR really fixes the mem leaks problems and more importantly, what is the guideline for future code to not create leaks? Also, are we able to create automated test for it?

**Comment by farmaazon** [comment]
> These changes look OK, but I'm missing info how this PR really works. Even if we are cloning world and scene in different places, all of these objects should get dropped if we drop the main app instance. Even if FRP closures keep reference to them, these networks should be dropped, so how this PR really fixes the mem leaks problems and more importantly, what is the guideline for future code to not create leaks? Also, are we able to create automated test for it?

There were two causes I'm aware of
1. `ApplicationData` FPR network was taking `ApplicationData` itself, making a reference cycle.
2. Our garbage collector: it often took FRP and model of various components, possibly containing reference to Application, or World. Of course, it should be removed a couple of frames later, but it's hard to say what exactly was dropped and what not (maybe handlers for GC were?).

Removing clones was rather for reducing noise when debugging (the reports said about 1500+ references to App), and I think removing reference which is not needed is a good idea anyway.

How to mitigate it? Here we have another example of FRP network taking itself. Almost all leaks I encounter in our codebase are like that. I think the FRP could be changed a bit, to make clear that FRP should own the model (maybe special "model" behavior node, which gives mutable references to Model?

**Comment by MichaelMauderer** [comment]
QA :yellow_circle: 

This looks mostly good, there are no errors after dropping the app. But when I press tab with focus in the empty window, I see the following error:
```
8pkg.js:3934 Uncaught Error: closure invoked after being dropped
    at imports.wbg.__wbindgen_throw (eval at <anonymous> (index.ts:296:37), <anonymous>:18716:11)
    at wasm_bindgen::throw_str::hd5a1e55c1ad9682d (pkg-opt.wasm:0x2179bdc)
    at <dyn core::ops::function::Fn<(A,)>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::hde034fa7641b5b81 (pkg-opt.wasm:0x216c679)
    at __wbg_adapter_80 (eval at <anonymous> (index.ts:296:37), <anonymous>:16975:8)
    at real (eval at <anonymous> (index.ts:296:37), <anonymous>:16936:14)
```
Could this indicate there is still some Enso event handling active? @farmaazon

**Comment by farmaazon** [comment]
> Could this indicate there is still some Enso event handling active? @farmaazon

Both yes and no: it means that the closure was dropped in rust successfully, but JS still had reference to it. See https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/closure/struct.Closure.html

Tracking the offending closure may be tricky, but I'll try to find it.

<blockquote><div><strong><a href="https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/closure/struct.Closure.html">Closure in wasm_bindgen::closure - Rust</a></strong></div><div>A handle to both a closure in Rust as well as JS closure which will invoke the Rust closure.</div></blockquote>

**Comment by farmaazon** [comment]
Added more serious tracking of `Scene` and `SceneData`, and it turns out there still are many leaks:
1. `init_webgl_2_context` is called with `Scene` as an argument, which is captured to lambdas, whose handlers are in turn kept inside `Scene`.
2. `CachedShapesPass` keeps scene while being also kept inside Renderer.
3. `font::Registry::init_and_load_embedded_fonts` clones scene and put it into structure returned as an extension - and the extensions are also kept inside scene
4. Finally, the scene is stored in `SCENE` global variable but never removed from there.

I still don't know why only `keydown` handlers raises an error, but those `Scene` leaks are serious.

**Comment by farmaazon** [comment]
Because the problems are not actually regressions (there were before this PR) and this PR improves things, I will merge it, but won't close the task #6505.

**Pull request merged**